retention_rule.go

```diff
@@ -0,0 +1,290 @@
+// each retention period
+type RetentionRule interface {
+	// RetentionPolicies are the retention policies at the time of the rule, ordered
+	// by retention period (from shorted to longest)
+	RetentionPolicies() []RetentionPolicy
+	SetRetentionPolicies(p []RetentionPolicy) RetentionRule
+
+	// CutoverTime is the time that the rule will begin to be applied
+	CutoverTime() time.Time
+	SetCutoverTime(t time.Time) RetentionRule
+
+	// CutoffTime is the time that the rule no longer applies
+	CutoffTime() time.Time
+	SetCutoffTime(t time.Time) RetentionRule
+}
+
+// NewRetentionRule creates a new retention rule
+func NewRetentionRule() RetentionRule { return new(retentionRule) }
+
+// RetentionRulesByCutoffTime is a sort.Interface for sorting RetentionRules by
+// CutoffTime, with the latest cutoff time first
+type RetentionRulesByCutoffTime []RetentionRule
+
+// Less compares two retention rules by their cutoff time
+func (rr RetentionRulesByCutoffTime) Less(i, j int) bool {
+	if rr[i].CutoffTime().IsZero() {
+		return true
+	}
+
+	if rr[j].CutoffTime().IsZero() {
+		return false
+	}
+
+	return rr[i].CutoffTime().After(rr[j].CutoffTime())
+}
+
+// Swap swaps two retention rules in the slice
+func (rr RetentionRulesByCutoffTime) Swap(i, j int) { rr[i], rr[j] = rr[j], rr[i] }
+
+// Len returns the length of the retention rule slice
+func (rr RetentionRulesByCutoffTime) Len() int { return len(rr) }
+
+// RetentionRuleProvider looks up retention rules for a given id and timerange
+type RetentionRuleProvider interface {
+	// FindRetentionRules returns the list of retention rules that apply
+	// for the given id over the requested time range
+	FindRetentionRules(id string, start, end time.Time) ([]RetentionRule, error)
+}
+
+type retentionPolicy struct {
+	r Resolution
+	p RetentionPeriod
+}
+
+func (p retentionPolicy) Resolution() Resolution           { return p.r }
+func (p retentionPolicy) RetentionPeriod() RetentionPeriod { return p.p }
+func (p retentionPolicy) Equal(other RetentionPolicy) bool {
+	return p.r.Equal(other.Resolution()) && p.p.Equal(other.RetentionPeriod())
+}
+
+type retentionRule struct {
+	policies                []RetentionPolicy
+	cutoffTime, cutoverTime time.Time
+}
+
+func (r *retentionRule) RetentionPolicies() []RetentionPolicy { return r.policies }
+func (r *retentionRule) CutoverTime() time.Time               { return r.cutoverTime }
+func (r *retentionRule) CutoffTime() time.Time                { return r.cutoffTime }
+
+func (r *retentionRule) SetRetentionPolicies(p []RetentionPolicy) RetentionRule {
+	r.policies = p
+	return r
+}
+
+func (r *retentionRule) SetCutoverTime(t time.Time) RetentionRule {
+	r.cutoverTime = t
+	return r
+}
+
+func (r *retentionRule) SetCutoffTime(t time.Time) RetentionRule {
+	r.cutoffTime = t
+	return r
+}
+
+type query struct {
+	xtime.Range
+	RetentionPolicy
+}
+
+type retentionQueryPlanner struct {
+	log   xlog.Logger
+	clock clock.Clock
```

Comment:
The assumption is that the rules are already sorted by the time they enter ``buildRetentionQueryPlan``.  Idea being that we'd load the rules from the config service once, sort them, then cache them in sorted order and simply re-use them for every query execution.